### DISCUSSION TITLE
Is there a way to disable `unsafe` option in `remark-stringify`?

### DISCUSSION THREAD

**Discussion by iAdramelk** [remarkjs/remark#648] [Q&A] [answered]
Basically I try to use `remark-lint` on mdx v2 files. And mdx v2 have uses `{/* comment */}` syntax for comments. Then `remark` enconters this stricture it replace it with `{/\* comment \*/}` which breaks build.

Is there a way to disable `unsafe` either globally, or just for this rule?

**Comment by wooorm** [comment, marked as answer]
there’s a lot going on here.

1. No, you should not turn off unsafe, that’s going to break *everything*
2. MDX and MD are different things
3. You‘re asking about remark-lint but the rest of your post ignores it

**Reply by iAdramelk** [reply]
Sorry, I asked XY question. Let me rephrase my question and add some context.

I want to lint and automatically fix my mdx files. I use mdx v2 in them so I need to use `{/* comment */}` syntax.

I tried following approaches:

- `prettier` - breaks comments (replaces `{/* comment*/}` with `{/_ comment _/}`) and some other syntax and can't be configured.
- `eslint-mdx` in `--fix` mode – makes a bunch of very strange changes and make code unparsable then some of the `remark-lint` rules are added (e.g. `lint-no-consecutive-blank-lines` adds bunch of `\\\\\\` in some places).
- `remark-cli` & `remark-link` with `-o` flag and options. Works for linting, but again, breaks comments as I wrote in the example above because of `unsafe`.

Is there a way to correctly lint and fix mdx v2 files?

**Reply by wooorm** [reply]
mdx@2 is very beta. So prettier hasn’t caught up (and I find that understandable).

It should work okay-ish with remark-cli if `remark-mdx` is added.

For eslint-mdx, that sounds like problems you might want to report there.

**Reply by iAdramelk** [reply]
Thanks for the answer.

> It should work okay-ish with remark-cli if remark-mdx is added.

Where should I add it? My `.remarkrc.js` looks like this:

```
module.exports = {
  settings: {
    bullet: "-",
    ruleRepetition: 3,
    fences: true,
    incrementListMarker: true,
    checkBlanks: true,
    resourceLink: true,
    emphasis: "*",
    tablePipeAlign: false,
    tableCellPadding: true,
    listItemIndent: 1,
  },
  plugins: [
    "frontmatter3",
    "mdx",
    "lint-no-file-name-mixed-case",
    "lint-no-file-name-articles",
    "lint-no-file-name-irregular-characters",
    "lint-no-file-name-consecutive-dashes",
    "lint-no-file-name-outer-dashes",
    "lint-no-consecutive-blank-lines",
    "lint-hard-break-spaces",
    ["lint-heading-style", "atx"],
    "lint-heading-increment",
    "lint-no-multiple-toplevel-headings",
    ["lint-no-heading-punctuation", ":."],
    ["lint-blockquote-indentation", 2],
    "lint-no-blockquote-without-marker",
    ["lint-unordered-list-marker-style", "-"],
    ["lint-ordered-list-marker-style", "."],
    ["lint-ordered-list-marker-value"],
    ["lint-list-item-indent", "space"],
    "lint-list-item-content-indent",
    ["lint-code-block-style", "fenced"],
    ["lint-fenced-code-flag", { allowEmpty: true }],
    ["lint-fenced-code-marker", "`"],
    ["lint-rule-style", "---"],
    "lint-no-table-indentation",
    "lint-table-pipes",
    ["lint-table-pipe-alignment", false],
    "lint-no-inline-padding",
    "lint-no-shortcut-reference-image",
    "lint-final-definition",
    "lint-definition-case",
    "lint-definition-spacing",
    ["lint-link-title-style", '"'],
    ["lint-strong-marker", "*"],
    ["lint-emphasis-marker", "*"],
    "lint-no-literal-urls",
    "lint-no-auto-link-without-protocol",
  ],
};
```

And the I run it from cli with the `-o` flag it still escapes comments.

**Reply by wooorm** [reply]
Use the remark-mdx that’s also part of the mdx@2 beta

**Reply by wooorm** [reply]
As in, I’m guessing that you just install v1, and not the latest v2 beta

**Reply by ChristianMurphy** [reply]
also note that in mdx 1

```jsx
{/* comment */}
```

is not considered a comment,
it would need to be in a jsx block, like so:

```jsx
<>{/* comment */}</>
```

see: https://astexplorer.net/#/gist/a6bdfb623bd185ef7a339ef991995e34/ddfd3dc41ab31b0a9a99b8146fdfbc919afbd96c

this is why prettier behaves strangely.
This may be resolved in MDX 2, @wooorm could confirm.

**Reply by iAdramelk** [reply]
I'm using v2 here. Just checked my `package.json`: `"remark-mdx": "^2.0.0-next.8",`.

**Reply by iAdramelk** [reply]
Checked `yarn.lock`, only `2.0.0-next.8` is installed. No older versions installed by some other lib.

**Reply by wooorm** [reply]
Oh, darn... They annoying part here is that mdx@2, latest beta, is for the previous versions of remark.
I’ve rewritten everything in mdx to support the new remark, but that was never released as an mdx beta.

:/

**Reply by iAdramelk** [reply]
If I use `remalk-cli` v8 instead of v9, will it fix it?

**Reply by wooorm** [reply]
yeah, but also other stuff has mostly updated, so that’d need to be downgraded too. Everything except MDX.

**Reply by iAdramelk** [reply]
Tried it actually. It works for the most part. But some rules are now broken:

- `no-consecutive-blank-lines` - now triggers on every node after jsx one. For some reason it assumes that jsx component is a bunch of black lines.
- `table-pipes` - It throws `Missing initial pipe in table fence` on tables inside jsx components.

Will disable these rules for now. It's still better that linting and fixing everything manually. Thanks a lot for the help.

**Reply by wooorm** [reply]
Pfew. Sorry you had so much trouble.
I'm personally also a bit annoyed that mdx is moving a bit slow, so I made wooorm/xdm as a modern alternative. It won't help you here though.
Also: would appreciate separate issues (and hopefully PRs?!) for these bugs

**Reply by iAdramelk** [reply]
No problem, I'm using beta software so its normal for some thing not yet be updated. :) Will create tickets for now.

**Reply by iAdramelk** [reply]
Added few tickets to `remark-lint` with some info about reasons they happens. Not sure what is the best what to fix them through.

**Comment by ChristianMurphy** [comment]
As @wooorm notes mdx and markdown are different.
It sounds like you're looking for something along the lines of https://github.com/mdx-js/eslint-mdx#mdxremark which can lint both the jsx and the markdown aspects for mdx?

**Reply by iAdramelk** [reply]
Thanks for the answer. I wrote about `eslint-mdx` in this comment https://github.com/remarkjs/remark/discussions/648#discussioncomment-426619 . It didn't quite work for me.
